### ISSUE TITLE
multiplication at the end of class_balanced_sigmoid_cross_entropy

### ISSUE THREAD

**Issue by dralves** [ppwwyyxx/tensorpack#278]
Hi
  This is mostly a question, but I couldn't find a mailing list so here it goes.
  At the end of [tensorpack.tfutils.symbolic_functions.class_balanced_sigmoid_cross_entropy](https://github.com/ppwwyyxx/tensorpack/blob/master/tensorpack/tfutils/symbolic_functions.py#L94) there is a multiplication of the loss tensor by the batch's positive ratio.
  However in the non-logit (and non-tf) implementation above the aforementioned method there's a multiplication of the positive's losses with the positives ratio _and_ a multiplication of the negative's losses with the negative ratio, this is consistent with the authors original [code](https://github.com/s9xie/hed/blob/9e74dd710773d8d8a469ad905c76f4a7fa08f945/src/caffe/layers/sigmoid_cross_entropy_loss_layer.cpp#L32)
  This additional multiplication makes this function have a different output that normal cross entropy when the positive ratio is 0.5, which was unexpected for me.
  Can you please give some insight into that factor? Is it because only the positives were scaled up? i.e the ratio is now 1:pos_weight and multiplying by the positive ratio would bring the positive factor back to 1? shouldn't this be done only for the positive labels?

**Comment by ppwwyyxx** [comment]
I don't fully understand your question. From what I can tell the two implementation handle the ratio in the same way.
Maybe you didn't see this line of code? ` cost = tf.reduce_mean(cost * (1 - beta))`

**Comment by dralves** [comment]
Sorry if I wasn't clear, I guess my wondering is whether that line should instead be something like:
`cost = tf.reduce_mean(tf.where(tf.equal(y_true, 1.), cost * (1 - beta), cost))`

**Comment by ppwwyyxx** [comment]
Because that would be wrong I think.
The test of whether label is true or false is already done in weighted_cross_entropy. Pos/Neg cost are already weighted.

If you think the current code is not equivalent to the original formula please show why.

**Comment by ppwwyyxx** [comment]
If that's what's confusing you: `pos_weight` in my code is __not__ the positive ratio, 1-beta is.

Closing now. Please reopen if you still have questions.
